algebra/euclidean_domain.lean

```diff
@@ -0,0 +1,324 @@
+            ... = (a/a)*a + a%a - (a/a)*a : by dsimp [d]; refl
+            ... = a - (a/a)*a : by dsimp [(%),(/)]; rw [witness a a],
+    rw this,
+    exact dvd_sub (dvd_refl a) (dvd_mul_left a (a/a)),
+end
+
+lemma mod_lt [decidable_euclidean_domain α]  :
+                     ∀ (a : α) {b : α}, valuation b > valuation (0:α) →  valuation (a%b) < valuation b :=
+begin
+    intros a b h,
+    cases decidable.em (b=0) with h',
+    {
+        rw h' at h,
+        have := lt_irrefl (valuation (0:α)),
+        contradiction,
+    },
+    {
+        cases val_mod a b with h h',
+        { contradiction },
+        { exact h' }
+    }
+end
+
+lemma neq_zero_lt_mod_lt [decidable_euclidean_domain α]  (a b : α) :  b ≠ 0 → valuation (a%b) < valuation b :=
+begin
+    intro hnb,
+    cases val_mod a b,
+    { contradiction },
+    { exact h }
+end
+
+lemma dvd_mod [decidable_euclidean_domain α] {a b c : α} : c ∣ a → c ∣ b → c ∣ a % b :=
+begin
+    intros dvd_a dvd_b,
+    have : remainder a b = a - quotient a b * b, from
+    calc 
+        a%b = quotient a b * b + a%b - quotient a b * b : by ring
+        ... = a - quotient a b * b : by {dsimp[(%)]; rw (witness a b)},
+    dsimp [(%)], rw this,
+    exact dvd_sub dvd_a (dvd_mul_of_dvd_right dvd_b (a/b)),
+end
+
+/- strong lemmas -/
+
+lemma val_lt_one [decidable_euclidean_domain α] (a : α) : 
+valuation a < valuation (1:α) → a = 0 :=
+begin
+    intro a_lt,
+    cases val_mul (1:α) a,
+    { exact h },
+    {
+        simp at h,
+        have := not_le_of_lt a_lt,
+        contradiction
+    }
+end
+
+lemma val_dvd_le [decidable_euclidean_domain α] (a b : α) :
+    b ∣ a → a ≠ 0 → valuation b ≤ valuation a :=
+begin
+    intros b_dvd ha,
+    induction b_dvd with x hx, rw hx,
+    cases val_mul b x,
+    {
+        rw h at hx, simp at hx,
+        contradiction,
+    },
+    {
+        exact h,
+    }
+end
+
+@[simp] lemma mod_one [decidable_euclidean_domain α] (a : α) : a % 1 = 0 :=
+begin
+    cases val_mod a 1,
+    { 
+        have := one_ne_zero, 
+        contradiction,
+    },
+    {
+        exact val_lt_one (a % 1) h,
+    }
+end 
+
+@[simp] lemma zero_mod  [decidable_euclidean_domain α] (b : α) : 0 % b = 0 :=
+begin
+    have h1 := witness 0 b,
+    have h2 : remainder 0 b = b * (-quotient 0 b ), from
+        calc
+            remainder 0 b = quotient 0 b * b + remainder 0 b + b * (-quotient 0 b ) : by ring
+            ...                       = b * (-quotient 0 b ) : by rw [h1, zero_add],
+    cases val_mul b (-quotient 0 b),
+    {
+        simp at h, rw h at h1, simp at h1,
+        exact h1,
+    },
+    {
+        rw [←h2] at h,
+        cases val_mod 0 b,
+        {
+            rw h_1, exact mod_zero (0:α),
+        },
+        {
+            have := not_le_of_lt h_1,
+            contradiction,
+        }
+    }
+end
+
+@[simp] lemma zero_div [decidable_euclidean_domain α] (b : α) (hnb : b ≠ 0) : 0 / b = 0 :=
+begin
+    have h1 := zero_mod b, dsimp [(%)] at h1,
```

Comment:
nope, don't do this. 

state what `h1` is, and then enclose its proof using `{ exact zero_mod b, dsimp [(%)] }` or similar.